File: action/src/test/kotlin/org/virtuslab/bazelsteward/action/ReleaseTagResolverTest.kt
package org.virtuslab.bazelsteward.action

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test

class ReleaseTagResolverTest {
  private val taggedCommitSha = "15ba5fa2b7eb9d9f2e67edb8cb355130b96d7a4d"
  private val otherCommitSha = "cccccccccccccccccccccccccccccccccccccccc"
  private val fakeReleases = listOf(
    "v1.7.2-rc9\tRC",
    "v1.7.2\tRelease 1.7.2",
    "v1.7.2.1\tPatch",
    "v1.7.3\tRelease 1.7.3",
  )
  private val fakeTags = listOf(
    RepositoryTag("v1.7.2", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
    RepositoryTag("v1.7.2.1", taggedCommitSha),
    RepositoryTag("v1.7.3", "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
    RepositoryTag("v1.7.2-rc9", taggedCommitSha),
  )
  private val fakeGh = FakeMetadataProvider(
    releases = fakeReleases,
    tags = fakeTags,
    refs = mapOf("main" to taggedCommitSha, "develop" to otherCommitSha),
  )

  @Test
  fun `v prefix resolves to latest matching stable release`() {
    ReleaseTagResolver.resolveLatestMatchingRelease("v1.7", fakeReleases) shouldBe "v1.7.3"
  }

  @Test
  fun `exact v tag ref resolves to latest matching patch release`() {
    ReleaseTagResolver.resolveLatestMatchingRelease("v1.7.2", fakeReleases) shouldBe "v1.7.2.1"
  }

  @Test
  fun `commit SHA resolves to release tag pointing to the same commit`() {
    ReleaseTagResolver.resolve(
      taggedCommitSha,
      "VirtusLab/bazel-steward",
      fakeGh,
    ) shouldBe "v1.7.2.1"
  }

  @Test
  fun `branch ref resolves to release tag via branch head commit`() {
    ReleaseTagResolver.resolve("main", "VirtusLab/bazel-steward", fakeGh) shouldBe "v1.7.2.1"
  }

  @Test
  fun `unknown branch ref fails`() {
    shouldThrow<IllegalStateException> {
      ReleaseTagResolver.resolve("non-existing-branch", "VirtusLab/bazel-steward", fakeGh)
    }
  }

  @Test
  fun `commit SHA without matching release tag fails`() {
    shouldThrow<IllegalStateException> {
      ReleaseTagResolver.resolve(
        otherCommitSha,
        "VirtusLab/bazel-steward",
        fakeGh,
      )
    }
  }

  @Test
  fun `matchesTagPattern accepts optional numeric suffix segments`() {
    ReleaseTagResolver.matchesTagPattern("v1.7.2.1", "v1.7.2") shouldBe true
    ReleaseTagResolver.matchesTagPattern("v1.7.2-rc9", "v1.7.2") shouldBe false
    ReleaseTagResolver.matchesTagPattern("v1.7.3", "v1.7") shouldBe true
  }

  @Test
  fun `short commit SHA resolves to release tag`() {
    val shortSha = taggedCommitSha.take(12)
    ReleaseTagResolver.resolve(shortSha, "VirtusLab/bazel-steward", fakeGh) shouldBe "v1.7.2.1"
  }

  @Test
  fun `ref resolving to commit without release tag fails`() {
    shouldThrow<IllegalStateException> {
      ReleaseTagResolver.resolve("develop", "VirtusLab/bazel-steward", fakeGh)
    }
  }

  @Test
  fun `commit SHA can resolve to rc release tag`() {
    val rcOnlyGh = FakeMetadataProvider(
      releases = listOf("v1.7.2-rc9\tRC"),
      tags = listOf(RepositoryTag("v1.7.2-rc9", taggedCommitSha)),
      refs = mapOf("main" to taggedCommitSha),
    )
    ReleaseTagResolver.resolve(
      taggedCommitSha,
      "VirtusLab/bazel-steward",
      rcOnlyGh,
    ) shouldBe "v1.7.2-rc9"
  }

  private class FakeMetadataProvider(
    private val releases: List<String>,
    private val tags: List<RepositoryTag>,
    private val refs: Map<String, String>,
  ) : GhReleaseMetadataProvider {
    override fun listReleases(repository: String): List<String> = releases

    override fun listTags(repository: String): List<RepositoryTag> = tags

    override fun resolveRefToCommitSha(repository: String, ref: String): String? = refs[ref]
  }
}

File: action/src/test/python/release_tag_resolver_test.py
import unittest

from action.release_tag_resolver import ReleaseTagResolver, RepositoryTag


class FakeMetadataProvider:
    def __init__(self, releases, tags, refs):
        self._releases = releases
        self._tags = tags
        self._refs = refs

    def list_releases(self, repository):
        return self._releases

    def list_tags(self, repository):
        return self._tags

    def resolve_ref_to_commit_sha(self, repository, ref):
        return self._refs.get(ref)


class ReleaseTagResolverTest(unittest.TestCase):
    tagged_commit_sha = "15ba5fa2b7eb9d9f2e67edb8cb355130b96d7a4d"
    other_commit_sha = "cccccccccccccccccccccccccccccccccccccccc"
    fake_releases = [
        "v1.7.2-rc9",
        "v1.7.2",
        "v1.7.2.1",
        "v1.7.3",
    ]
    fake_tags = [
        RepositoryTag("v1.7.2", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
        RepositoryTag("v1.7.2.1", tagged_commit_sha),
        RepositoryTag("v1.7.3", "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
        RepositoryTag("v1.7.2-rc9", tagged_commit_sha),
    ]

    def make_provider(self):
        return FakeMetadataProvider(
            releases=self.fake_releases,
            tags=self.fake_tags,
            refs={"main": self.tagged_commit_sha, "develop": self.other_commit_sha},
        )

    def test_v_prefix_resolves_to_latest_matching_stable_release(self):
        self.assertEqual(
            ReleaseTagResolver.resolve_latest_matching_release("v1.7", self.fake_releases),
            "v1.7.3",
        )

    def test_exact_v_tag_ref_resolves_to_latest_matching_patch_release(self):
        self.assertEqual(
            ReleaseTagResolver.resolve_latest_matching_release("v1.7.2", self.fake_releases),
            "v1.7.2.1",
        )

    def test_commit_sha_resolves_to_release_tag_pointing_to_the_same_commit(self):
        self.assertEqual(
            ReleaseTagResolver.resolve(self.tagged_commit_sha, "VirtusLab/bazel-steward", self.make_provider()),
            "v1.7.2.1",
        )

    def test_branch_ref_resolves_to_release_tag_via_branch_head_commit(self):
        self.assertEqual(
            ReleaseTagResolver.resolve("main", "VirtusLab/bazel-steward", self.make_provider()),
            "v1.7.2.1",
        )

    def test_unknown_branch_ref_fails(self):
        with self.assertRaisesRegex(
            RuntimeError,
            "Could not resolve ref non-existing-branch in VirtusLab/bazel-steward to a commit SHA",
        ):
            ReleaseTagResolver.resolve("non-existing-branch", "VirtusLab/bazel-steward", self.make_provider())

    def test_commit_sha_without_matching_release_tag_fails(self):
        with self.assertRaisesRegex(
            RuntimeError,
            (
                "No GitHub release tag in VirtusLab/bazel-steward points to commit "
                "cccccccccccccccccccccccccccccccccccccccc \\(resolved from "
                "cccccccccccccccccccccccccccccccccccccccc\\)"
            ),
        ):
            ReleaseTagResolver.resolve(self.other_commit_sha, "VirtusLab/bazel-steward", self.make_provider())

    def test_matches_tag_pattern_accepts_optional_numeric_suffix_segments(self):
        self.assertTrue(ReleaseTagResolver.matches_tag_pattern("v1.7.2.1", "v1.7.2"))
        self.assertFalse(ReleaseTagResolver.matches_tag_pattern("v1.7.2-rc9", "v1.7.2"))
        self.assertTrue(ReleaseTagResolver.matches_tag_pattern("v1.7.3", "v1.7"))

    def test_human_readable_gh_release_table_line_is_not_treated_as_release_tag(self):
        self.assertIsNone(
            ReleaseTagResolver.extract_release_tag("Release 1.7.2\tLatest\tv1.7.2\tabout 1 day ago"),
        )

    def test_short_commit_sha_resolves_to_release_tag(self):
        short_sha = self.tagged_commit_sha[:12]
        self.assertEqual(
            ReleaseTagResolver.resolve(short_sha, "VirtusLab/bazel-steward", self.make_provider()),
            "v1.7.2.1",
        )

    def test_ref_resolving_to_commit_without_release_tag_fails(self):
        with self.assertRaisesRegex(
            RuntimeError,
            (
                "No GitHub release tag in VirtusLab/bazel-steward points to commit "
                "cccccccccccccccccccccccccccccccccccccccc \\(resolved from develop\\)"
            ),
        ):
            ReleaseTagResolver.resolve("develop", "VirtusLab/bazel-steward", self.make_provider())

    def test_commit_sha_can_resolve_to_rc_release_tag(self):
        rc_only_provider = FakeMetadataProvider(
            releases=["v1.7.2-rc9"],
            tags=[RepositoryTag("v1.7.2-rc9", self.tagged_commit_sha)],
            refs={"main": self.tagged_commit_sha},
        )
        self.assertEqual(
            ReleaseTagResolver.resolve(self.tagged_commit_sha, "VirtusLab/bazel-steward", rc_only_provider),
            "v1.7.2-rc9",
        )


if __name__ == "__main__":
    unittest.main()
